src/Ogooreck.Sample.BusinessLogic.FSharp.Tests/Deciders/BankAccountDecider.fs

```diff
@@ -0,0 +1,80 @@
+﻿module Deciders.BankAccountDecider
+
+open System
+open BankAccount
+open BankAccountPrimitives
+
+type OpenBankAccount =
+    { BankAccountId: AccountId
+      AccountNumber: AccountNumber
+      ClientId: ClientId
+      CurrencyISOCode: CurrencyCode
+      Now: DateTimeOffset }
+
+type RecordDeposit = { Amount: decimal; CashierId: CashierId; Now: DateTimeOffset }
+type WithdrawCashFromATM = { Amount: decimal; AtmId: AtmId; Now: DateTimeOffset }
+type CloseBankAccount = { Reason: string; Now: DateTimeOffset }
+
+type Command =
+    | OpenBankAccount of OpenBankAccount
+    | RecordDeposit of RecordDeposit
+    | WithdrawCashFromATM of WithdrawCashFromATM
+    | CloseBankAccount of CloseBankAccount
+
+let openBankAccount (command: OpenBankAccount) (bankAccount: BankAccount) =
+    match bankAccount with
+    | Open _ -> invalidOp "Account is already opened!"
+    | Closed _ -> invalidOp "Account is already closed!"
+    | NotInitialised _ ->
+        BankAccountOpened
+            { BankAccountId = command.BankAccountId
+              AccountNumber = command.AccountNumber
+              ClientId = command.ClientId
+              CurrencyISOCode = command.CurrencyISOCode
+              CreatedAt = command.Now
+              Version = 1 }
+
+let recordDeposit (command: RecordDeposit) (bankAccount: BankAccount) =
+    match bankAccount with
+    | NotInitialised _ -> invalidOp "Account is not opened!"
+    | Closed _ -> invalidOp "Account is closed!"
+    | Open openBankAccount ->
+        DepositRecorded
+            { BankAccountId = openBankAccount.Id
+              Amount = command.Amount
+              CashierId = command.CashierId
+              RecordedAt  = command.Now
+              Version = openBankAccount.Version + 1L }
+
+let withdrawCashFromATM (command: WithdrawCashFromATM) (bankAccount: BankAccount) =
+    match bankAccount with
+    | NotInitialised _ -> invalidOp "Account is not opened!"
+    | Closed _ -> invalidOp "Account is closed!"
+    | Open openBankAccount ->
+        if (openBankAccount.Balance < command.Amount) then
+            invalidOp "Not enough money!"
+
+        CashWithdrawnFromATM
+            { BankAccountId = openBankAccount.Id
+              Amount = command.Amount
+              ATMId = command.AtmId
+              RecordedAt = command.Now
+              Version = openBankAccount.Version + 1L }
+
+let closeBankAccount (command: CloseBankAccount) (bankAccount: BankAccount) =
```

Comment:
Type annotations probably not required.

Reply:
Yeah, I'm still anxious about using type interference everywhere. From my past experience, I prefer to name explicit types where I know that I don't expect to have any other type ever. Thoughts?

Reply:
Legibility should be the main driver - if you think mentioning the type will give a useful cue to a reader, thats a good reason to put it in. If you have 50 in a row the types, parens and colons make a mess.

Pinning the type can sometimes even show up in compile perf (only with pretty large assemblies though - this is not anything to worry about).

The other reason to pin even when not strictly necessary is if something is used in lots of contexts and is low in the stack - sometimes a bad inference can happen based on the first usage which then ripples as lots of compiler errors from all the other usages.